### DISCUSSION TITLE
Why are are item actions not supporting await?

### DISCUSSION THREAD

**Discussion by Tr3ffel** [directus/directus#11046] [Archive]
Currently I want to change the response payload inside an action handler:
```js
const afterRead = async ({ payload }) => {
    payload.forEach((item) => {
      item.test = 'test';
    });
  }
};
action('items.items.read', afterRead);
```
This works as expected. 

But in my case I need to query the database and than update the field based on the query.
After quick testing I discovered that actions are called asynchronously (non-blocking).

```js
const wait = new Promise((res) => {
  setTimeout(() => res(true), 1000);
});

const afterRead = async ({ payload }) => {
    await wait();
    payload.forEach((item) => {
      item.test= 'test';
    });
};
action('items.items.read', afterRead);

// item.test will be undefined in response
```

Why are items actions not supporting await?
In my option this would be a common usecase.

**Comment by oreilles** [comment]
Actions are run asynchronously, they're rather intended to making external calls or running long task in the background without blocking execution. Furthermore, [as stated in the docs](https://docs.directus.io/extensions/hooks/#action) they're run *after* the event has actually ocurred (in your case, the item would have already been sent by the API when your code would be executed, meaning it wouldn't have change the query result.

If you need to update the payload with an async function, you want to use [`filter`](https://docs.directus.io/extensions/hooks/#filter) instead.
